### PR TITLE
Backend: User login

### PR DESCRIPTION
**Pull request by lpil**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -5,6 +5,8 @@ gem 'rails', '~> 5.0.0'
 gem 'pg', '~> 0.18'
 gem 'puma', '~> 3.0'
 gem 'bcrypt', '~> 3.1.7'
+gem 'jwt'
+gem 'has_secure_token'
```
Comment (lpil) [inline Gemfile]: I think this is built into Rails 5 now. Do we need this extra lib?

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -20,6 +21,6 @@ def create
   private
 
   def user_params
-    params.fetch(:user, {}).permit(:email, :password)
+    params.fetch(:user, {}).permit(:email, :password, :token)
```
Comment (lpil) [inline app/controllers/users_controller.rb]: We shouldn't take the token from the user here. They give us email and password, we give them the user.

File: notes.md [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 
-users should be able to add their tea and pronoun preferences
+```sh
+rails s
+```
+To start and run postgres database via docker:
 
-users should be able to view other user's tea and pronoun preferences
+```sh
+docker run postgres
```
Comment (lpil) [inline notes.md]: This shouldn't be in the repo, it's entirely specific to the user's machine. Each dev will do this differently.

Comment (lpil) [inline notes.md]: Oh, this is your personal notes. Bit odd that it's in repo, but OK. I thought it was the README, oops :)

File: spec/controllers/users_controller_spec.rb [hunk 4]
```diff
@@ -24,7 +24,7 @@
         }
       }
       expect(response.status).to eq 201
-      expect(response.body).to be_json_eql({
+      expect(response.body).to include_json({
```
Comment (lpil) [inline spec/controllers/users_controller_spec.rb]: Lets keep using `be_json_eql`, but also add the token key to the object.

We don't know what it is, but we can look it up with `User.first.token`

**Comment by lpil** [comment]
@scripttease Looks good! I've left a few little comments.

File: notes.md (hunk 3 above)
Comment (scripttease) [inline notes.md]: i tried to put the notes in .gitignore but somehow messed that up :/

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -1,25 +1,34 @@
...
+        }
+      }
     end
   end
 
   private
 
   def user_params
-    params.fetch(:user, {}).permit(:email, :password, :token)
+    params.fetch(:user, {}).permit(:email, :password)
```
Comment (lpil) [inline app/controllers/sessions_controller.rb]: This is correct, but we don't actually use it anywhere :)

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+sudo: required
```
Comment (lpil) [inline .travis.yml]: Why is sudo required?

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+language: ruby
+
+rvm:
+  - 2.2.2
+
+services:
+  - docker
```
Comment (lpil) [inline .travis.yml]: Shouldn't need docker, only postgres. They probably already have postgres running :)

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+before_script:
+  - docker run postgres
```
Comment (lpil) [inline .travis.yml]: No need for this either, they have their own way of running postgres.

Remember that the way that postgres is run is entirely down to the user/environment. We just happen to be using docker, no one else has to.

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
+class SessionsController < ApplicationController
+  def create
+    user = User.find_by(email: params[:user][:email])
+    if user && user.authenticate(params[:user][:password])
+      # session[:user] = user
```
Comment (lpil) [inline app/controllers/sessions_controller.rb]: Delete this line please :)

File: config/routes.rb [hunk 10]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
   post "/users" => "users#create"
+  post "/login" => "sessions#create"
+  # delete "/logout" => "sessions#destroy"
```
Comment (lpil) [inline config/routes.rb]: Delete this comment please :)

**Comment by lpil** [comment]
This is great. Nearly there :D

File: .travis.yml (hunk 7 above)
Comment (scripttease) [inline .travis.yml]: ok that will definitely help it not crash then :)
thanks

File: spec/controllers/sessions_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+      )
+      post :create, params: {
+        user: {
+          email: "hello@world.com",
+        }
+      }
+      expect(response.status).to eq 400
+    end
+
+    xit "does not create a session if no params given" do
```
Comment (lpil) [inline spec/controllers/sessions_controller_spec.rb]: Pending tests here. Get it working :)

**Pull request merged**